homeassistant/components/shelly/__init__.py

```diff
@@ -249,7 +249,13 @@ def _async_rpc_device_setup() -> None:
             )
             platforms = RPC_PLATFORMS
 
-        hass.config_entries.async_setup_platforms(entry, platforms)
+        await hass.config_entries.async_forward_entry_setups(entry, platforms)
+
+        # TODO: only connect the scanner if enabled in config options active/true/false
+        if not sleep_period:
+            entry.async_on_unload(
```

Comment:
I think we can move this bellow line 250

Reply:
Will likely need to tweak it a bit since we only want to turn it on if the shelly is up and running since we need to pull the scanner if we loose the connection to the shelly so bluetooth knows it has gone away and will immediately switch to the next nearest one for a given device.

For that we need to know when it gets disconnected.  I couldn't find a way to get callbacks when the connection is established or disconnected so I think that probably will have to be a change in `aioshelly`

Reply:
Actually if we reached this stage the device is already connected, else we will get an error raised here:
https://github.com/home-assistant/core/blob/f17041e239f243324801076a814a4d487aede946/homeassistant/components/shelly/__init__.py#L276

We had a disconnect event which I removed since it was not used, I can add it back. We still know if device is disconnected if it doesn't send an event within a time period:
https://github.com/home-assistant/core/blob/f17041e239f243324801076a814a4d487aede946/homeassistant/components/shelly/coordinator.py#L450

Reply:
> We had a disconnect event which I removed since it was not used, I can add it back. We still know if device is disconnected if it doesn't send an event within a time period:

That would be great.

The goal is to cancel the scanner when we get disconnected via the callback returned from `async_connect_scanner`

https://github.com/home-assistant/core/pull/82007/files#diff-ef0eaf3053c447ec8f2cbfd32a151528d6eb6a9d012d10a9e86e73e706c7bbe5R47

...and than when we reconnect, connect the scanner back up by calling `async_connect_scanner` again